File: CPP/DP/45. Jump Game II.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int jumps = 0, curEnd = 0, curReach = 0, goal = nums.size()-1;
        for (int i = 0; i <= goal; i++){
            curReach = max(curReach, i + nums[i]);
            if (curEnd == goal) return jumps;
            if (i == curEnd) {
                jumps++;
                curEnd = curReach;
            }
        }
        return jumps;
    }
};

File: Python/DP/45. Jump Game II.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        cur_reach = 0
        next_max_reach = 0
        jumps = 0

        for i in range(n):
            if i > cur_reach:
                jumps += 1
                cur_reach = next_max_reach
            next_max_reach = max(i + nums[i], next_max_reach)
        
        return jumps